keps/sig-node/3960-pod-lifecycle-sleep-action/README.md

```diff
@@ -0,0 +1,436 @@
+
+##### e2e tests
+- Basic functionality
+  1. Create a simple pod with a container that runs a long-running process.
+  2. Add a preStop hook to the container configuration, using the new sleepAction with a specified sleep duration (e.g., 5 seconds).
+  3. Delete the pod and observe the time it takes for the container to terminate.
+  4. Verify that the container sleeps for the specified duration before it is terminated.
+  5. Verify that the container keeps executing code while sleeping.
+
+- Sleep duration boundary testing
+  1. Create a simple pod with a container that runs a long-running process.
+  2. Add a preStop hook to the container configuration, using the new sleepAction with various sleep durations, including:1 seconds (minimum allowed value), values slightly above the minimum allowed value (to test edge cases).
+  3. For each sleep duration, delete the pod and observe the time it takes for the container to terminate.
+  4. Verify that the container sleeps for the specified duration before it is terminated.
+  5. Verify that the container keeps executing code while sleeping.
+
+- Container exit/crash testing
+  1. Create a simple pod with a container that will exit soon.
+  2. Add a preStop hook to the container configuration, using the new sleepAction with a specified sleep duration longer than the container's lifecycle.
+  3. Exit the container and observe the time it takes for the pod to terminate.
+  4. Verify that the pod be deleted successfully without waiting for the entire sleep duration.
+
+- Interaction with termination grace period
+  1. Create a simple pod with a container that runs a long-running process.
+  2. Add a preStop hook to the container configuration, using the new sleepAction with a specified sleep duration (e.g., 5 seconds).
+  3. Set the termination grace period to various values, including:
+     - Equal to the sleep duration
+     - Greater than the sleep duration
+     - Greater than the sleep duration, but reduced to less than the sleep duration at runtime
+  4. For each termination grace period value, delete the pod and observe the time it takes for the container to terminate.
+  5. Verify that the container is terminated after the min(sleep, grace).
+
+### Graduation Criteria
+
+#### Alpha
+
+- Feature implemented behind a feature flag
+- Initial unit/e2e tests completed and enabled
+- Documentation is added to demonstrate why this is useful in nginx scenario and how exactly nginx needs to be configured.
+
+#### Beta
+
+- Gather feedback from developers and surveys
+- Additional e2e tests are completed(if needed)
+
+#### GA
+
+- No negative feedback
+- No bug issues reported
+
+### Upgrade / Downgrade Strategy
+
+#### Upgrade
+The previous PreStop behavior will not be broken. Users can continue to use their hooks as it is.
+To use this enhancement, users need to enable the feature gate, and add sleep action in their prestop hook.
+
+#### Downgrade
+Kube-apiserver will ignore sleep in prestop hook.
```

Comment:
Let's be more precise:
1) It will silently drop the sleep prestop hook if someone wants to add that to the object (or create object with it)
2) However, existing objects that have it set will not be cleared.

[This also touches on the comment I see from @jeremyrickard below.]